### PR TITLE
Specify custom URL for Twilio

### LINKED ISSUE #878: Add base url setting to twilio channels
### 🎯 The task

De-hardcode the Twilio base URL. Let the user set it for every channel.

### 🥕 The motivation

The main reasons are:

- Twilio can have multiple endpoints. Having an only URL for all Twilio channels prohibits users from taking advantage of that.
- Surveda (instedd/surveda#1793) needs an IVR stress test tool and it could take the shape of a fake Twilio endpoint that uses the new setting.

### 🚧 Tasks:

- [x] Add base_url to twilio channel
- [x] When blank use the default `https://api.twilio.com/2010-04-01`
- [x] Add it in rails forms
- [x] Pick up from erlang the base_url of the current twilio channel
- [x] (Optional) add it in pigeon screens (used as iframes for Surveda)

### PR DESCRIPTION
**Pull request by ysbaddaden**
Reverts the revert + adds a fix to the erlang broker that should hopefully be working (to be verified).

related to #879 
closes #878 
closes https://github.com/instedd/surveda/issues/2226

### CODE REVIEW

**Comment by ysbaddaden** [comment]
Hum, the `Dockerfile` seems broken :thinking:

File: app/models/channels/twilio.rb [hunk 1]
```diff
@@ -19,11 +19,14 @@ class Channels::Twilio < Channel
   config_accessor :account_sid
   config_accessor :auth_token
   config_accessor :number
+  config_accessor :base_url
 
   attr_protected :guid
 
   before_create :create_guid
 
+  validates :base_url, format: URI::regexp(["http", "https"]), allow_nil: true, allow_blank: true
```
Comment (matiasgarciaisaia) [inline app/models/channels/twilio.rb:28]: If there's an option to allow nil (ie, default URL) but not blanks (ie, reject a blank non-default URL), I'd go for it?

File: app/views/channels_ui/_form_twilio.haml [hunk 2]
```diff
@@ -9,3 +9,6 @@
 
 .row
   = f.text_field :limit, colspan: 's4', helper: 'Number of concurrent calls'
+
+.row
+  = f.text_field :base_url, label: 'Twilio base URL'
```
Comment (matiasgarciaisaia) [inline app/views/channels_ui/_form_twilio.haml:14]:
```suggestion
  = f.text_field :base_url, label: 'Custom Twilio base URL'
```

Comment (matiasgarciaisaia) [reply app/views/channels_ui/_form_twilio.haml:14]: Adding what's the current default may help users tell what format the base URL should have (something like `Leave empty to default to https://api.twilio.com/2010-04-01` - so users know they don't need to add the trailing forward-slash, but they do need to include the version of the API).

**Review by matiasgarciaisaia — APPROVED** [review]
LGTM! 👍

**Comment by ysbaddaden** [comment]
@matiasgarciaisaia updated with the default URL that also appears as placeholders:

---
**From Surveda:**
Capture d’écran de 2023-03-27 19-01-53@2x

---
**From Verboice:**
Capture d’écran de 2023-03-27 19-02-16@2x

**Comment by matiasgarciaisaia** [comment]
:shipit:

File: spec/models/channels/twilio_spec.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    it "accepts url" do
+      channel = Channels::Twilio.make_unsaved(base_url: "http://domain:8000/resource")
+      expect(channel.valid?).to eq(true)
+      expect(channel.base_url).to eq("http://domain:8000/resource")
+    end
+
+    it "saves blank string as nil" do
+      channel = Channels::Twilio.make_unsaved(base_url: "  \t  ")
+      expect(channel.valid?).to eq(true)
+      expect(channel.base_url).to eq(nil)
```
Comment (matiasgarciaisaia) [inline spec/models/channels/twilio_spec.rb:37]: Woah, I didn't get from the code how this test case works. I was about to suggest changing the validation to support this one, but I'm glad it works!

**Review by matiasgarciaisaia — APPROVED** [review]
👍

**Pull request merged**